xarray/core/dataset.py

```diff
@@ -2748,6 +2748,64 @@ def roll(self, **shifts):
 
         return self._replace_vars_and_dims(variables)
 
+    def sortby(self, variables, ascending=True):
+        """
+        Sorts the dataset, either along specified dimensions,
+        or according to values of 1-D dataarrays that share dimension
+        with calling object.
+
+        Parameters
+        ----------
+        variables: (str, DataArray, or iterable of either)
+            Name of a 1D variable in coords/data_vars whose values are used to
+            to sort the dataset.
+        ascending: boolean, optional.
+            whether to sort by ascending order.
+
+        Returns
+        -------
+        sorted: Dataset
+            A new dataset where all the specified dims are sorted by dim
+            labels.
+        """
+        from .dataarray import DataArray
+        if isinstance(variables, (str, unicode, DataArray)):
```

Comment:
To reduce the number of separate code paths, I would suggest a normalization based flow. For example:
```python
if not isinstance(variables, list):
    variables = [variables]
# variables is now a list of str/DataArray
variables = [v if isinstance(v, DataArray) else self[v]
             for v in variables]
# variables is now a list of DataArray
```
Now you can avoid the `if`/`else` branch inside the loop.